text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1121 @@
+# CDK Construct Upgrade Support
+
+- **Original Author(s):**: @GavinZZ
+- **Tracking Issue**: #715
+- **API Bar Raiser**: @iliapolo
+
+A construct upgrade framework in the CDK CLI that helps developers upgrade from deployed and
+older construct libraries to modern, fully supported alternatives that incorporate the latest
+features and best practices.
+
+## Motivation
+
+As the AWS Cloud Development Kit (CDK) evolves, we continuously deliver new features,
+performance optimizations, and reliability improvements. To maintain backward compatibility
+while enabling forward progress, the CDK team uses structured versioning strategies for critical updates.
+
+A common pattern involves releasing versioned constructs or modules (e.g., V2 variants)
+that coexist with their predecessors. For example:
+
+* The `@aws-cdk/aws-ec2-alpha` alpha module introduces the modernized `VpcV2`, `SubnetV2`, and other
+new construct, designed to succeed the original Vpc related constructs in `aws-cdk-lib/aws-ec2`.
+* The `TableV2` construct in `aws-cdk-lib/aws-dynamodb` replaces legacy Table implementations
+while retaining API compatibility.
+
+These updates unlock enhanced capabilities. Construct upgrade between variants, however, often requires
+code changes that could trigger CloudFormation resource replacements (e.g., NAT gateways being
```

Comment:
Internally with our own wrapper CDK we are discussing the idea of a semantic versioning schema of four entries X.X.X.X where you have major infrastructure deployment breaking change, major code breaking change, minor, patch. This allows us to decouple refactors from live service breaking change. Especially in a world with code assistants making breaking library code changes far less disruptive if non-breaking to the live resource. Just wanted to throw this out there as an idea to discuss for the aws-cdk team